cpp/src/parquet/encoding.cc

```diff
@@ -2537,6 +2537,100 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
 // ----------------------------------------------------------------------
 // DELTA_LENGTH_BYTE_ARRAY
 
+// ----------------------------------------------------------------------
+// DeltaLengthByteArrayEncoder
+
+class DeltaLengthByteArrayEncoder : public EncoderImpl,
+                                    virtual public TypedEncoder<ByteArrayType> {
+ public:
+  using T = typename ByteArrayType::c_type;
+
+  explicit DeltaLengthByteArrayEncoder(const ColumnDescriptor* descr, MemoryPool* pool)
+      : EncoderImpl(descr, Encoding::DELTA_LENGTH_BYTE_ARRAY,
+                    pool = ::arrow::default_memory_pool()),
+        sink_(pool),
+        length_encoder_(nullptr, pool),
+        encoded_size_{0},
+        lengths_{0} {}
+
+  std::shared_ptr<Buffer> FlushValues() override;
+
+  int64_t EstimatedDataEncodedSize() override { return encoded_size_; }
+
+  using TypedEncoder<ByteArrayType>::Put;
+
+  void Put(const ::arrow::Array& values) override;
+
+  void Put(const T* buffer, int num_values) override;
+
+  void PutSpaced(const T* src, int num_values, const uint8_t* valid_bits,
+                 int64_t valid_bits_offset) override;
+
+ protected:
+  ::arrow::BufferBuilder sink_;
+  DeltaBitPackEncoder<Int32Type> length_encoder_;
+  uint32_t encoded_size_;
+  std::vector<int32_t> lengths_;
+};
+
+void DeltaLengthByteArrayEncoder::Put(const ::arrow::Array& values) {
+  auto src = values.data()->GetValues<ByteArray>(1);
+  Put(src, static_cast<int>(values.length()));
+}
+
+void DeltaLengthByteArrayEncoder::Put(const T* src, int num_values) {
+  if (num_values == 0) {
+    return;
+  }
+  lengths_.resize(num_values);
```

Comment:
It seems that `lengths_` vector is not required and thus can be removed.

Reply:
Perhaps we should `lengths_.resize(0)` to keep it small between `Put` calls?

Reply:
"Vector capacity is never reduced when resizing to smaller size"
https://en.cppreference.com/w/cpp/container/vector/resize

Also, I'm not sure what you're trying to achieve. What is the point of the temporary lengths vector here?

Reply:
@mapleFU [originally proposed](https://github.com/apache/arrow/pull/14293#discussion_r1050922988) to use `DeltaLengthByteArrayEncoder.lengths_` between `Put` calls to avoid the vector initialization cost. I am not sure of the of the impact either way, but resizing to 0 is then not an option. Shall we then revert to a local `lengths_` variable?

Reply:
Let's stare back a bit. You don't _need_ a lengths variable at all.

Reply:
Do you mean something like:
```cpp
void DeltaLengthByteArrayEncoder::Put(const T* src, int num_values) {
  [...]
  for (int idx = 0; idx < num_values; idx++) {
    const int32_t len = src[idx].len;
    length_encoder_.Put(&len, 1);
    encoded_size_ += len;
  }

  for (int idx = 0; idx < num_values; idx++) {
    sink_.UnsafeAppend(src[idx].ptr, src[idx].len);
  }
}
```

Reply:
> Let's stare back a bit. You don't _need_ a lengths variable at all.

Sorry for do the misleading. I just want to make clear our code style. But I think here are two ways:

method1:

```
for every bytes:
   (temp variable) length = bytes.length
   Put(&length, 1)
```

or  method2

```
Copy all lengths to a vector
Put(vec.data(), vec.size());
```

Seems that we prefer the 

Seems, When should we buffer the vector values, like ? And when should we just create a vector? Seems there would be a `sizeof(variable) * BatchSize` of memory allocation.

By the way, I found there are both `ArrowPoolVector` (with arrow allocator) and `std::vector` (with std allocator), when should we use them? @pitrou

Reply:
Yes... but now I realize that putting values one by one to the `length_encoder_` might be quite suboptimal.
The best compromise might be to use a small stack-allocated buffer, e.g. (untested):
```c++
void DeltaLengthByteArrayEncoder::Put(const T* src, int num_values) {
  [...]
  constexpr int kBatchSize = 256;
  std::array<int32_t, kBatchSize> lengths;
  for (int idx = 0; idx < num_values; idx += kBatchSize) {
    const int batch_size = std::min(kBatchSize, num_values - idx);
    for (int j = 0; j < batch_size; ++j) {
      const int32_t len = src[idx].len;
      encoded_size_ += len;
      lengths[j] = len;
    }
    length_encoder_.Put(lengths.data(), batch_size);
  }
  [...]
```

Reply:
Yeah, that makes sense. [Changed](https://github.com/apache/arrow/pull/14293/commits/219e7d74670717f3f89a4df703eb194ae400b15a).